[inductor][fx passes] batch layernom

jackiexu1992 opened the pull request:
Summary: Batch layernorm. Fuse independent horizontal layernorm with same size into one.

Test Plan:
# unit test
```
buck test mode/dev-nosan //caffe2/test/inductor:group_batch_fusion
File changed: fbcode//caffe2/test/inductor/test_group_batch_fusion.py
Buck UI: https://www.internalfb.com/buck2/68eb51e1-bdbc-4847-aabf-e50737d8485b
Test UI: https://www.internalfb.com/intern/testinfra/testrun/5066549764442206
Network: Up: 0 B  Down: 0 B
Jobs completed: 10. Time elapsed: 1:07.2s.
Tests finished: Pass 3. Fail 0. Fatal 0. Skip 0. Build failure 0
```

Differential Revision: D47447542



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

jackiexu1992 commented:
@jansel @yanboliang for review.

jackiexu1992 commented:
easycla

jansel commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -156,6 +156,78 @@ def fuse(self, graph, subset):
+                "batch_layernorm",
+                str(input_node.meta["example_value"].shape)
+                if "example_value" in input_node.meta
+                else "",
```
What happens with missing example values?  Should this prevent fusion?

jackiexu1992 replied:
Make sense. Should prevent such fusions.

jansel commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -156,6 +156,78 @@ def fuse(self, graph, subset):
+            group_inputs.append(get_arg_value(node, 0, "input"))
+            group_shapes.append(get_arg_value(node, 1, "normalized_shape"))
+            group_weights.append(get_arg_value(node, 2, "weight"))
+            group_biases.append(get_arg_value(node, 3, "bias"))
```
`bias` is an optional arg.  What happens if some have bias set and others don't?

jansel commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -156,6 +156,78 @@ def fuse(self, graph, subset):
+            group_nodes.append(node)
+            group_inputs.append(get_arg_value(node, 0, "input"))
+            group_shapes.append(get_arg_value(node, 1, "normalized_shape"))
+            group_weights.append(get_arg_value(node, 2, "weight"))
```
`weight` is an optional arg.   What happens if some have weight set and others don't?

jansel commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -156,6 +156,78 @@ def fuse(self, graph, subset):
+            if eps is None:
+                eps = 1e-5
+            group_epss.append(eps)
+        stack_dim = -1 - len(group_shapes[-1])
```
Add asserts that each item in these lists matches.

jackiexu1992 commented:
Comments resolved :) cc @jansel

jansel commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -156,6 +156,116 @@ def fuse(self, graph, subset):
+                    str(input.meta["example_value"].shape),
+                    str(weight.meta["example_value"].shape) if weight is not None else "",
+                    str(get_arg_value(node, 1, "normalized_shape")),
+                    weight is None,
+                    bias is None,
```
Do these need to be shapes too?  

Not sure if things like broadcasting are legal.

jackiexu77 committed:
[inductor][fx passes] batch layernom (#105492)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/105492

Batch layernorm. Fuse independent horizontal layernorm with same size into one.

Test Plan:
# unit test
```
buck test mode/dev-nosan //caffe2/test/inductor:group_batch_fusion
File changed: fbcode//caffe2/torch/_inductor/fx_passes/group_batch_fusion.py
File changed: fbsource//xplat/caffe2/torch/_inductor/fx_passes/group_batch_fusion.py
Buck UI: https://www.internalfb.com/buck2/b22f635a-ffef-4ea8-8222-217a11a4f689
Test UI: https://www.internalfb.com/intern/testinfra/testrun/562950274863832
Network: Up: 0 B  Down: 0 B
Jobs completed: 10. Time elapsed: 2:10.8s.
Tests finished: Pass 3. Fail 0. Fatal 0. Skip 0. Build failure 0
```

Reviewed By: jansel, xuzhao9

Differential Revision: D47447542

fbshipit-source-id: 6ee425703408f767207e671af602de6bd58458b4
